src/main/java/io/cdap/http/NettyHttpService.java

```diff
@@ -301,7 +306,63 @@ public Thread newThread(Runnable r) {
       executor.setKeepAliveTime(execThreadKeepAliveSecs, TimeUnit.SECONDS);
       executor.allowCoreThreadTimeOut(true);
     }
-    return new NonStickyEventExecutorGroup(executor);
+
+    // Returns a EventExecutorGroup that overrides the inEventLoop behavior of the EventExecutor returned by the group.
+    // The inEventLoop will return true if it is calling from a thread in the same thread group, hence was created
+    // from the same executor.
+    // This is to make sure channel events coming from a thread owned by the EventExecutor
+    // will be executed in the same thread instead of submitting a new task.
+    return new ForwardingEventExecutorGroup(new NonStickyEventExecutorGroup(executor)) {
```

Comment:
Any reason not changing NonStickyEventExecutorGroup's inEventLoop behavior ? just because of avoiding any behavior change?

Reply:
the NonStickyEventExecutorGroup should comes from the netty library. We copied here to fix a race condition bug, which we intended to remove the class after upgrading netty to a newer version.